[windows] Avoid %r for quoting module-cache-path in Windows.

drodriguez opened the pull request:
%r returns a representation of the object that is valid Python syntax.
For numbers and strings this representation is very compatible with
Unix shells, but for Windows, the quoting performed by Python will not
be compatible. For example the Windows path separator `\` will be
escaped as `\\`, which is not necessary for Windows and might make
some tests that try to match path fail.

Python 3.3 has `shlex.quote`, which was previously available as
`pipes.quote` in earlier Python versions. `pipes.quote` was indeed
used in several points of the `lit.cfg` file. For Windows, one need to
do their own quoting.

This change introduces `shell_quote` which uses `shlex.quote` or
`pipes.quote` in Unix depending on the Python version, and provides an
implementation of `shell_quote` for Windows. It replaces every usage
of `pipes.quotes` for `shell_quote`, and modifies the value of
`mcp_opt` to use `shell_quote` and not `%r`.

This should fix the test `Driver\working-directory.swift` in the
Windows Visual Studio 2017 builder.

drodriguez committed:
[windows] Avoid %r for quoting module-cache-path in Windows.

%r returns a representation of the object that is valid Python syntax.
For numbers and strings this representation is very compatible with
Unix shells, but for Windows, the quoting performed by Python will not
be compatible. For example the Windows path separator `\` will be
escaped as `\\`, which is not necessary for Windows and might make
some tests that try to match path fail.

Python 3.3 has `shlex.quote`, which was previously available as
`pipes.quote` in earlier Python versions. `pipes.quote` was indeed
used in several points of the `lit.cfg` file. For Windows, one need to
do their own quoting.

This change introduces `shell_quote` which uses `shlex.quote` or
`pipes.quote` in Unix depending on the Python version, and provides an
implementation of `shell_quote` for Windows. It replaces every usage
of `pipes.quotes` for `shell_quote`, and modifies the value of
`mcp_opt` to use `shell_quote` and not `%r`.

This should fix the test `Driver\working-directory.swift` in the
Windows Visual Studio 2017 builder.

drodriguez commented:
@swift-ci please smoke test

drodriguez commented:
I might replace every usage of `%r` at some point, but for now, let's introduce only one change at the time.

compnerd reviewed (approved):
The windows path makes me sad :-(.  Thank you for cleaning this up.  Removing the `%r` usage sounds like a good idea to me.

compnerd commented:
Seems that `re.replace` doesn't work with the windows builders?

drodriguez commented:
Sorry. I run it in several of the test directories, but not all of them. It should have been `s.replace(…)` not `re.replace(…)`.

@swift-ci please smoke test

drodriguez commented:
@swift-ci please test Windows platform

drodriguez commented:
The problem now is that because the `\` are not longer escaped. There are usages of Lit's `SubstituteCaptures` (https://github.com/apple/llvm-project/blob/1a6659f0e7add3f1c1f5352fbf5b9811bac1adfc/llvm/utils/lit/lit/TestingConfig.py#L167) that use string interpolation with the non-escaped values. Because those strings end up as replacement strings of a regular expression, they have to be escaped for the replacement string (easy to do with `.replace('\\', '\\\\')`).

But then the `working-directory.swift` test starts to fail for some reason (FML). It seems that the matching is performed against some output of the Swift driver, which outputs the jobs command lines escaped, but probably escaped for Unix. I have to circle back here. The fix for one particular test might be easier than this, even if this should be good in the long run.

I don't know, however, if the card castle of escaping and unescaping can ever be solved.

PD: I still don't understand why it works in Windows VS2019.

drodriguez commented:
So I think I understand why it works in the VS 2019 nodes: they have Python 3.7.5 installed (comes from Visual Studio from reading the logs), while the VS 2017 nodes have 3.8.5 (installed with Chocolatey).

For `os.path.realpath` the documentation says "Changed in version 3.8: Symbolic links and junctions are now resolved on Windows.". When LLVM Lit starts it does a `realpath` on the lit configuration path which gives `T:\swift\test-windows-x86_64` in Python 3.7.5, but `C:\jenkins\workspace\oss-swift-windows-x86_64\build\swift\test-windows-x86_64` in Python 3.8.5.

And then enter `PathSanitizingFileCheck` into the game: it provides both `SOURCE_DIR` and `BUILD_DIR`, but `BUILD_DIR` value comes from CMake, which doesn't do any `realpath` shenanigans, so when Python 3.8.5 uses paths like `C:\jenkins\workspace\oss-swift-windows-x86_64\build\swift\test-windows-x86_64` as the build directory, `PathSanitizingFileCheck` still uses `T:\swift` as prefix, and do not matches both.

To put the cherry on top, if one uses `realpath` for the substitutions of `BUILD_DIR`, `PathSanitizingFileCheck` stops working because those paths where always using forward slashes, and `realpath` changes everything to forward slashes.

Since this problem will start appearing when the VS 2019 nodes update to Python 3.8.5, I don't think the solution is going back to 3.7.5 in the VS 2017 nodes, but the solution is not going to be pretty.

drodriguez commented:
@swift-ci please test Windows platform

drodriguez commented:
@swift-ci please smoke test

drodriguez commented:
@swift-ci please test Windows platform

drodriguez commented:
@swift-ci please smoke test

compnerd commented:
`os.normpath` should convert the path to the canonical spelling for the platform - that is `/` will be converted to `\` which might be what needs to be applied here?

drodriguez commented:
Do not merge yet. The Windows CI failed badly. I have to look and figure out what's happening in that machine (it is quite hard to get things right because I have to copy code back and forth between machines).

compnerd commented:
@swift-ci please test Windows platform

drodriguez commented:
Removed the part that adds carets. Seems that Lit already does some massaging of the arguments, and the quoting I introduced was misbehaving with the one Lit implements. I tested having a space in my Python path, like it seems to happen in the build machine.

@swift-ci please test Windows platform

drodriguez commented:
@swift-ci please smoke test

drodriguez commented:
@swift-ci please smoke test Linux platform

drodriguez commented:
@swift-ci please test Linux platform

drodriguez commented:
@swift-ci please test Linux platform
